rafthttp/snapshot_hub.go

```diff
@@ -0,0 +1,170 @@
+	if s.hasSnap {
+		plog.Panicf("unexpected snapshot register when previous one has not finished")
+	}
+	s.hasSnap = true
+	s.snapTo = to
+	s.snapIndex = index
+	s.snapClose = time.AfterFunc(snapshotWaitTimeout, func() {
+		s.mu.Lock()
+		defer s.mu.Unlock()
+		if !s.hasSnap {
+			return
+		}
+		s.hasSnap = false
+		// close the snapshot in the store
+		s.snapst.WriteSnapTo(&errWriter{errors.New("cancelled")})
+		plog.Infof("snapshot [index: %d, to: %s] failed to be sent out due to timeout", index, s.snapTo)
+		s.r.ReportSnapshot(uint64(to), raft.SnapshotFailure)
+	})
+}
+
+// writeSnapTo writes the snapshot with the given index into the given writer.
+func (s *snapshotHub) writeSnapTo(w io.Writer, index uint64) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if !s.hasSnap {
+		return fmt.Errorf("failed to find available snapshot")
+	}
+	if s.snapIndex != index {
+		return fmt.Errorf("failed to find wanted snapshot")
+	}
+	s.hasSnap = false
+	s.snapClose.Stop()
+
+	err := s.snapst.WriteSnapTo(w)
+	if err != nil {
+		plog.Infof("snapshot [index: %d, to: %s] failed to be sent out (%v)", index, s.snapTo, err)
+		s.r.ReportSnapshot(uint64(s.snapTo), raft.SnapshotFailure)
+	} else {
+		plog.Infof("snapshot [index: %d, to: %s] succeeded to be sent out", index, s.snapTo)
+		s.r.ReportSnapshot(uint64(s.snapTo), raft.SnapshotFinish)
+	}
+	return err
+}
+
+// processSnap processes the given raft snapshot message.
+// First, it sends requests to get full snapshot for raft snapshot message. Then
+// it lets snapshot store read out the snapshot. After the full snapshot is
+// received, it lets raft process the raft snapshot message.
+func (s *snapshotHub) processSnap(m raftpb.Message) {
+	remote := types.ID(m.From)
+	index := m.Snapshot.Metadata.Index
+	snapr, err := s.requestSnap(remote, index)
+	if err != nil {
+		plog.Warningf("failed to request snapshot [index: %d] (%v)", m.Snapshot.Metadata.Index, err)
+		return
+	}
+	defer snapr.Close()
+
+	if err := s.snapst.ReadSnapFrom(snapr, index); err != nil {
+		plog.Warningf("failed to read snapshot [index: %d] (%v)", m.Snapshot.Metadata.Index, err)
+		return
+	}
+	if err := s.r.Process(context.TODO(), m); err != nil {
+		plog.Warningf("failed to process raft message (%v)", err)
+	}
+}
+
+func (s *snapshotHub) requestSnap(remote types.ID, index uint64) (io.ReadCloser, error) {
+	up, ok := s.urlPickers[remote]
+	if !ok {
+		return nil, fmt.Errorf("cannot find url picker")
+	}
+	u := up.pick()
+	uu := u
+	uu.Path = RaftSnapPrefix
+	req, err := http.NewRequest("GET", uu.String(), nil)
+	if err != nil {
+		up.unreachable(u)
+		return nil, err
+	}
+	req.Header.Set("X-Snapshot-Index", strconv.FormatUint(index, 10))
+	resp, err := s.tr.RoundTrip(req)
+	if err != nil {
+		up.unreachable(u)
+		return nil, err
+	}
+
+	switch resp.StatusCode {
+	case http.StatusOK:
+		return resp.Body, nil
+	default:
+		resp.Body.Close()
+		return nil, fmt.Errorf("unexpected http status %s", http.StatusText(resp.StatusCode))
```

Comment:
Wait... I totally understand it is because user provided a wrong one. And we SHOULD log the error. However my question was `shall we also mark unreachable in this case?`.